File: lib/models-dev.ts
// Models.dev API client for frontend

const MODELS_DEV_API_URL = "https://models.dev/api.json";
const CACHE_TTL_MS = 60 * 60 * 1000; // 1 hour

let cache: { data: ModelsDevData; timestamp: number } | null = null;

export interface ModelsDevProvider {
  id: string;
  name: string;
  api?: string;
  env: string[];
  npm?: string;
  doc?: string;
  models: Record<string, ModelsDevModel>;
}

export interface ModelsDevModel {
  id: string;
  name: string;
  family?: string;
  attachment?: boolean;
  reasoning?: boolean;
  tool_call?: boolean;
  temperature?: boolean;
  modalities?: Modalities;
  cost?: Cost;
  limit?: Limit;
  open_weights?: boolean;
  knowledge?: string;
  release_date?: string;
  last_updated?: string;
}

export interface Modalities {
  input: string[];
  output: string[];
}

export interface Cost {
  input: number;
  output: number;
  cache_read?: number;
  cache_write?: number;
}

export interface Limit {
  context: number;
  output: number;
}

export interface ModelsDevData {
  providers: Record<string, ModelsDevProvider>;
}

export function supportsVision(model: ModelsDevModel): boolean {
  return model.attachment === true || (model.modalities?.input?.includes("image") ?? false);
}

export function supportsText(model: ModelsDevModel): boolean {
  return model.modalities?.output?.includes("text") ?? false;
}

export function formatCost(cost: Cost): string {
  if (cost.input === 0 && cost.output === 0) return "Free";
  return `$${cost.input}/${cost.output} per 1M tokens`;
}

export function formatContext(limit: Limit): string {
  if (limit.context >= 1000000) return `${(limit.context / 1000000).toFixed(1)}M`;
  if (limit.context >= 1000) return `${(limit.context / 1000).toFixed(0)}K`;
  return `${limit.context}`;
}

/// Fetch models.dev data from API or cache
export async function fetchModels(): Promise<ModelsDevData> {
  // Check cache
  if (cache && Date.now() - cache.timestamp < CACHE_TTL_MS) {
    return cache.data;
  }

  const response = await fetch(MODELS_DEV_API_URL);
  if (!response.ok) {
    throw new Error(`Failed to fetch models.dev: ${response.status}`);
  }

  const raw: Record<string, ModelsDevProvider> = await response.json();
  const data: ModelsDevData = { providers: raw };

  // Update cache
  cache = { data, timestamp: Date.now() };

  return data;
}

/// Get all providers with API URLs (usable for OpenAI-compatible calls)
export async function listProviders(): Promise<ModelsDevProvider[]> {
  const data = await fetchModels();
  return Object.values(data.providers).filter((p) => p.api);
}

/// Get all models from all providers with API URLs
export async function listAllModels(): Promise<{ provider: ModelsDevProvider; model: ModelsDevModel }[]> {
  const providers = await listProviders();
  return providers.flatMap((provider) =>
    Object.values(provider.models).map((model) => ({ provider, model }))
  );
}

/// Get models suitable for vision (image text extraction)
export async function listVisionModels(): Promise<{ provider: ModelsDevProvider; model: ModelsDevModel }[]> {
  const all = await listAllModels();
  return all.filter(({ model }) => supportsVision(model));
}

/// Get models suitable for text (metadata extraction)
export async function listTextModels(): Promise<{ provider: ModelsDevProvider; model: ModelsDevModel }[]> {
  const all = await listAllModels();
  return all.filter(({ model }) => supportsText(model));
}

/// Get a model by its full ID (provider/model)
export async function getModelById(fullId: string): Promise<{ provider: ModelsDevProvider; model: ModelsDevModel } | null> {
  const data = await fetchModels();

  // Split on first '/' to get provider and model
  const slashIndex = fullId.indexOf("/");
  if (slashIndex === -1) return null;

  const providerId = fullId.slice(0, slashIndex);
  const modelId = fullId.slice(slashIndex + 1);

  const provider = data.providers[providerId];
  if (!provider) return null;

  const model = provider.models[modelId];
  if (!model) return null;

  return { provider, model };
}

/// Get a model, searching only in configured providers
export async function getModelFromConfiguredProviders(
  modelId: string,
  configuredProviderIds: string[]
): Promise<{ provider: ModelsDevProvider; model: ModelsDevModel } | null> {
  // First try as "provider/model" format
  const result = await getModelById(modelId);
  if (result) return result;

  // Search in configured providers
  const data = await fetchModels();
  for (const providerId of configuredProviderIds) {
    const provider = data.providers[providerId];
    if (provider) {
      const model = provider.models[modelId];
      if (model) {
        return { provider, model };
      }
    }
  }

  return null;
}

/// Clear the cache
export function clearCache(): void {
  cache = null;
}

File: src-tauri/src/models/models_dev.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct Modalities {
    #[serde(default)]
    pub input: Vec<String>,
    #[serde(default)]
    pub output: Vec<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct Cost {
    #[serde(default)]
    pub input: f64,
    #[serde(default)]
    pub output: f64,
    #[serde(default)]
    pub cache_read: Option<f64>,
    #[serde(default)]
    pub cache_write: Option<f64>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
pub struct Limit {
    #[serde(default)]
    pub context: u64,
    #[serde(default)]
    pub output: u64,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ModelsDevModel {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub family: Option<String>,
    #[serde(default)]
    pub attachment: bool,
    #[serde(default)]
    pub reasoning: bool,
    #[serde(default)]
    pub tool_call: bool,
    #[serde(default)]
    pub temperature: bool,
    #[serde(default)]
    pub modalities: Modalities,
    #[serde(default)]
    pub cost: Cost,
    #[serde(default)]
    pub limit: Limit,
    #[serde(default)]
    pub open_weights: bool,
    #[serde(default)]
    pub knowledge: Option<String>,
    #[serde(default)]
    pub release_date: Option<String>,
    #[serde(default)]
    pub last_updated: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ModelsDevProvider {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub api: Option<String>,
    #[serde(default)]
    pub env: Vec<String>,
    #[serde(default)]
    pub npm: Option<String>,
    #[serde(default)]
    pub doc: Option<String>,
    pub models: HashMap<String, ModelsDevModel>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ModelsDevData {
    pub providers: HashMap<String, ModelsDevProvider>,
}

impl ModelsDevData {
    pub fn from_json(json: &str) -> Result<Self, String> {
        let raw: HashMap<String, ModelsDevProvider> = serde_json::from_str(json)
            .map_err(|e| format!("Failed to parse models.dev JSON: {}", e))?;
        Ok(Self { providers: raw })
    }

    pub fn get_model(&self, full_id: &str) -> Option<(ModelsDevProvider, ModelsDevModel)> {
        let slash_pos = full_id.find('/')?;
        let provider_id = &full_id[..slash_pos];
        let model_id = &full_id[slash_pos + 1..];

        let provider = self.providers.get(provider_id)?;
        let model = provider.models.get(model_id)?;
        Some((provider.clone(), model.clone()))
    }

    pub fn get_model_from_providers(
        &self,
        model_id: &str,
        provider_ids: &[String],
    ) -> Option<(ModelsDevProvider, ModelsDevModel)> {
        for provider_id in provider_ids {
            if let Some(provider) = self.providers.get(provider_id) {
                if let Some(model) = provider.models.get(model_id) {
                    return Some((provider.clone(), model.clone()));
                }
            }
        }
        None
    }
}
